### PR TITLE
Python pillar library

### PR DESCRIPTION
**Pull request by tacerus**
This imports the Python based pillar logic used in our Salt infrastructures in the form of reusable libraries, helping us to deduplicate code and making it easier to maintain changes. The respective "#!py" pillars will then be adjusted to reference library calls, with only the infrastructure specific data needing to be passed as function parameters.

### CODE REVIEW

File: infrastructure-formula/python/pillar/network.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <https://www.gnu.org/licenses/>.
+"""
+
+from logging import getLogger
+
+from yaml import safe_load
+
+root = '/srv/salt-git/pillar'
```
Comment (cboltz) [inline infrastructure-formula/python/pillar/network.py:22]: That's the third time I read this hardcoded path ;-)

Would it make sense to define it in a single place, to read it from the salt config (with possible issues if you define more than one salt root) or even to make it configurable?

File: infrastructure-formula/python/pillar/network.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+        ifconfig.pop(primary_interface, None)
+
+        log(f'{msg} primary IPv4 address set to {ip4}')
+        log(f'{msg} primary IPv6 address set to {ip6}')
+
+    # machine not in hosts, can be an error or expected (for example because the machine is bare metal or in an unsupported location)
+    else:
+        log(f'{msg} no host entry!')
+        return pillar
```
Comment (cboltz) [inline infrastructure-formula/python/pillar/network.py:96]: 70 lines above, you define `pillar = {}`, and it wasn't changed since then - so effectively this line means `return {}`.

Would it make sense to write it as `return {}` to make that clear, or do you have any plans to add something to `pillar` in the lines above?

(Similar question for line 30 - but at least that's closer to the definition.)

File: infrastructure-formulas.spec [hunk 3]
```diff
@@ -121,6 +123,14 @@ Requires:       %{name}-common
...
 Salt state for managing grains.
 
+%package -n hosts-formula
+Summary:        Salt states for managing %{_sysconfdir}/hosts
+License:        GPL-3.0-or-later
+Requires:       %{name}-common
+
+%description -n hosts-formula
+Salt states for managing the %{_sysconfdir}/hosts file.
+
```
Comment (cboltz) [inline infrastructure-formulas.spec:133]: hosts-formula looks unrelated to this PR - did you intentionally add it (and its %files)?

File: infrastructure-formulas.spec [hunk 4]
```diff
@@ -471,3 +503,13 @@ done
 %files -n zypper-formula -f zypper.files
 
 %changelog
+%files -n infrastructure-formula-python
```
Comment (cboltz) [inline infrastructure-formulas.spec:506]: The `%changelog` above is one too much.

File: infrastructure-formula/python/pillar/network.py (hunk 2 above)
Comment (cboltz) [reply infrastructure-formula/python/pillar/network.py:96]: Alternative idea: make this

```
if host not in domaindata['hosts']:
    return {}  # or   return pillar
```
and move it above line 45.

Then un-indent everything that currently lives inside `if host in domaindata['hosts']:` ;-)

File: infrastructure-formulas.spec (hunk 3 above)
Comment (tacerus) [reply infrastructure-formulas.spec:133]: I just run packaging/genspec.py.

Comment (tacerus) [reply infrastructure-formulas.spec:133]: I usually do that once for every release, but figured adding this Python functionality warrants an extraordinary run.

File: infrastructure-formulas.spec (hunk 4 above)
Comment (tacerus) [reply infrastructure-formulas.spec:506]: Correcting in the template.

File: infrastructure-formulas.spec (hunk 3 above)
Comment (tacerus) [reply infrastructure-formulas.spec:133]: I will drop it and have it part of the release commit as usual.

File: infrastructure-formula/python/README.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# Python pillar helpers
+
+A Python library to be used in `#!py` pillar SLS files. It allows for rendering of formula pillars based off data in YAML datasetes.
```
[the remark this answers is not in the document]
Comment (tacerus) [reply infrastructure-formula/python/README.md:3]: Thanks!

File: infrastructure-formula/python/pillar/network.py (hunk 1 above)
Comment (tacerus) [reply infrastructure-formula/python/pillar/network.py:22]: I was thinking to add a configuration file for these libraries, however it will require additional states to write it.
Alternative it can be passed using function arguments. Unfortunately I do not know of a way to access the pillar without passing it from function arguments.

Comment (tacerus) [reply infrastructure-formula/python/pillar/network.py:22]: However so far, the directory is the same in all our Salt environments, hence I think of it to be duplication to pass it in all function calls.

File: infrastructure-formula/python/pillar/network.py (hunk 2 above)
Comment (tacerus) [reply infrastructure-formula/python/pillar/network.py:96]: It's indeed equivalent. If you want I can define pillar only later, and make this `{}` explicit.

The early return with less indentation is a good idea, thanks.

File: infrastructure-formula/python/README.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+A Python library to be used in `#!py` pillar SLS files. It allows for rendering of formula pillars based off data in YAML datasets.
+The logic is opinionated and specific to the architecture of our code based infrastructure.
+
+## Usage
+
+
+Example pillar file `network.sls`:
+
+```
+
```
Comment (cboltz) [inline infrastructure-formula/python/README.md:12]: Nitpick: the blank lines 8 and 12 are superfluous

Comment (tacerus) [reply infrastructure-formula/python/README.md]: True, thanks

File: infrastructure-formula/python/pillar/network.py (hunk 2 above)
Comment (cboltz) [reply infrastructure-formula/python/pillar/network.py]: I tend to prefer ;-) `return {}` to make clear what's going on.

Comment (tacerus) [reply infrastructure-formula/python/pillar/network.py]: I think now the variable is clearly defined a few lines above

Comment (cboltz) [reply infrastructure-formula/python/pillar/network.py]: Looks good - thanks!

File: infrastructure-formula/python/pillar/network.py (hunk 1 above)
Comment (cboltz) [reply infrastructure-formula/python/pillar/network.py:22]: That sounds like making it configurable is not easy, and - at least for now - not worth the effort. So let's keep it hardcoded until we have a good reason to make it configurable.
